tapeagents/environment.py

```diff
@@ -203,3 +223,38 @@ def close(self) -> None:
 
     def last_actions(self, tape: Tape) -> list[Action]:
         return [step for step in tape.steps[-tape.metadata.n_added_steps :] if isinstance(step, Action)]
+
+    async def areact(self, tape: Tape) -> Tape:
+        for action in self.last_actions(tape):
+            t = time.perf_counter()
+            observation = await self.astep(action)
+            observation.metadata.other["action_execution_time"] = time.perf_counter() - t
+            observation.metadata.other["action_kind"] = action.kind
+            tape = tape.append(observation)
+        return tape
+
+    async def astep(self, action: Action) -> Observation:
+        action_type = type(action)
+        if isinstance(action, LLMOutputParsingFailureAction):
+            return UserStep(content="Try again")
+        if action_type not in self.action_map:
+            raise Exception(f"Unknown action: {action_type}")
+        tool = self.action_map[action_type]
+        if isinstance(tool, AsyncBaseTool):
+            observation = await tool.arun(action)
+        else:
+            logger.warning(f"Tool {tool} is not async and could slowdown the rollouts!")
+            observation = tool.run(action)
+        return observation
+
+    async def areset(self) -> None:
+        for tool in self.tools:
```

Comment:
If I am understanding this logic, it is implied that there should never be a mix of sync and async tools? Is it worth guarding against in the init?

Reply:
We can call sync tools from the async loop (with the warning about the potential slowdown) but not vice versa